awss3/store.go

```diff
@@ -0,0 +1,703 @@
+
+	q.Delimiter = "/"
+
+	// Think we should just put 1 here right?
+	itemLimit := int64(f.PageSize)
+	if q.PageSize > 0 {
+		itemLimit = int64(q.PageSize)
+	}
+
+	params := &s3.ListObjectsInput{
+		Bucket:    aws.String(f.bucket),
+		MaxKeys:   &itemLimit,
+		Prefix:    &q.Prefix,
+		Delimiter: &q.Delimiter,
+	}
+
+	folders := make([]string, 0)
+
+	for {
+		select {
+		case <-ctx.Done():
+			// If has been closed
+			return folders, ctx.Err()
+		default:
+			if q.Marker != "" {
+				params.Marker = &q.Marker
+			}
+			resp, err := f.client.ListObjectsWithContext(ctx, params)
+			if err != nil {
+				return nil, err
+			}
+			for _, cp := range resp.CommonPrefixes {
+				folders = append(folders, strings.Trim(*cp.Prefix, `/`))
+			}
+			return folders, nil
+		}
+	}
+}
+
+/*
+// Copy from src to destination
+func (f *FS) Copy(ctx context.Context, src, des cloudstorage.Object) error {
+
+	so, ok := src.(*object)
+	if !ok {
+		return fmt.Errorf("Copy source file expected s3 but got %T", src)
+	}
+	do, ok := des.(*object)
+	if !ok {
+		return fmt.Errorf("Copy destination expected s3 but got %T", des)
+	}
+
+	oh := so.b.Object(so.name)
+	dh := do.b.Object(do.name)
+
+	_, err := dh.CopierFrom(oh).Run(ctx)
+	return err
+}
+
+// Move which is a Copy & Delete
+func (f *FS) Move(ctx context.Context, src, des cloudstorage.Object) error {
+
+	so, ok := src.(*object)
+	if !ok {
+		return fmt.Errorf("Move source file expected s3 but got %T", src)
+	}
+	do, ok := des.(*object)
+	if !ok {
+		return fmt.Errorf("Move destination expected s3 but got %T", des)
+	}
+
+	oh := so.b.Object(so.name)
+	dh := do.b.Object(des.name)
+
+	if _, err := dh.CopierFrom(oh).Run(ctx); err != nil {
+		return err
+	}
+
+	return oh.Delete(ctx)
+}
+*/
+// NewReader create GCS file reader.
+func (f *FS) NewReader(o string) (io.ReadCloser, error) {
+	return f.NewReaderWithContext(context.Background(), o)
+}
+
+// NewReaderWithContext create new File reader with context.
+func (f *FS) NewReaderWithContext(ctx context.Context, objectname string) (io.ReadCloser, error) {
+	res, err := f.client.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Key:    aws.String(objectname),
+		Bucket: aws.String(f.bucket),
+	})
+	if err != nil {
+		// translate the string error to typed error
+		if strings.Contains(err.Error(), "NoSuchKey") {
+			return nil, cloudstorage.ErrObjectNotFound
+		}
+		return nil, err
+	}
+	return res.Body, nil
+}
+
+// NewWriter create Object Writer.
+func (f *FS) NewWriter(objectName string, metadata map[string]string) (io.WriteCloser, error) {
+	return f.NewWriterWithContext(context.Background(), objectName, metadata)
+}
+
+// NewWriterWithContext create writer with provided context and metadata.
+func (f *FS) NewWriterWithContext(ctx context.Context, objectName string, metadata map[string]string) (io.WriteCloser, error) {
+
+	// Create an uploader with the session and default options
+	uploader := s3manager.NewUploader(f.sess)
+
+	pr, pw := io.Pipe()
```

Comment:
nice use of `io.Pipe`, to convert this call into a WriterCloser!  

Note: https://golang.org/pkg/io/#Pipe:
```
That is, each Write to the PipeWriter blocks until it has satisfied one or more Reads from the PipeReader that fully consume the written data. The data is copied directly from the Write to the corresponding Read (or Reads); there is no internal buffering.
```
We may want to consider wrapping the write side of the pipe in a writer buffer: see https://github.com/lytics/cloudstorage/blob/master/csbufio/buf_writercloser.go#L17

I didn't know that with GCS because it already has an internal buffer.

Reply:
good idea, done